### ISSUE TITLE
support recursive update for submodules

### ISSUE THREAD

**Issue by shudson310** [oxygenxml/oxygen-git-plugin#64]
If a repo contains submodules, the current pull does not update the submodule. The plugin should support recursive update of submodules.

git submodule update --init --recursive

**Comment by AlexJitianu** [comment]
I agree. If the --recurse-submodule wasn't set when the repository was cloned, we should run `git submodule update --init --recursive` when the user tries to open a submodule.

**Comment by AlexJitianu** [comment]
There's a new build, 1.3.2 that will initialize/update the submodules whenever the user opens a submodule. Please give it try and let me know if this is indeed what you want.

**Comment by xephon2** [comment]
I think this is what Git users normally expect, how git submodules should work, but this slightly does not match the idea of Git submodules. Submodules are used to include dependencies and those should not always updated _automatically_. The correct way to implement this requirement (to update submodules on a pull) is to use the mentioned command above in a `post-merge` Git hook. The now implemented solution is probably what most of the users expect, but may lead to problems in edge cases.

**Comment by AlexJitianu** [comment]
Thanks for the insight, @xephon2 . If I understand this correctly... if you are using submodules then you have to create a post-merge Git hook yourself that will run the update command. That means that we actually didn't have to do anything on the plugin part, right? 

Anyway, if what we did is what Git users normally expect, then I will leave it like it is (initialize/update the submodules whenever the user opens a submodule) and see if anyone else has any objections.

**Comment by xephon2** [comment]
@AlexJitianu : To make sure that we are talking about the same things. It should be the default, that the plugin initializes all submodules recursively when _cloning_ a repository. But it _should not_ update the submodules, when the the user pulls new commits of the main repository or switches branches. The problem is, that Git submodules are pointers to commits, not branches. So, by default, if you switch a branch of the main repository and the HEAD commit of the branch points to an older commit of a submodule, your submodule would fall in a detached HEAD state, which will usually cause trouble, because commits will be lost, if they are made in detached HEAD state. But the user may want to keep the default Git behavior for any reasons. So the plugin should not change the way how Git works. So anything beside the inititialization of the submodules should be defined manually by the user by writing Git hooks to avoid changing the way how Git works.

**Comment by kirkilj** [comment]
So what was the final decision on this? We ask writers to associate a semver tag with a specific commit in a submodule so that version 3.4.6 of a parent repo can maintain the fact that it's only compatible with version 1.6.9 of the sub module.

Our writers have asked us to address the recursive update issue by instead notifying them if a submodule has received more pushed commits after the tagged version, so they can decide if they want to selectively update specific submodules. Basically provide a submodule status table with buttons for various actions, mainly update.

**Comment by AlexJitianu** [comment]
Hello,

I'm sorry for the delay. When working with submodules, I see the following cases for pulling changes.

### Case 1. Pulling in Upstream Changes from the Submodule Remote

Somebody committed in the submodule repository and you want the project that has a submodule to that repository to have those changes. There is no clear indicator now in the UI right now that the submodule repository has evolved.
To do this action in our Git Client, one has to :

- use the action on the toolbar to open the submodule.
- checkout the proper branch since the submodule is on a specific commit
- pull changes
- switch to the parent project and commit/push the submodule marker file.

This use case requires one to specify a branch from which to update. The user can either manually checkout the branch using the UI or the branch can be specified in a `.gitmodule` file:

```
[submodule "my/module"]
    path = my/module
    branch = master
```

In your case, I think the person doing this would be you. 

Since this is an operation that should be done with case, we don't want to automate it. What we want to do on the short term to help with this use case is to decorate the submodules action on the toolbar with pull information.
![ljjklckepkediodb](https://user-images.githubusercontent.com/7802649/107760884-25cec680-6d33-11eb-9c59-6f2214ac4e6d.png)
Please note that the tooltip will present useful information only if we know the submodule branch. Unless one uses a `.gitmodule` file to specify the branch, we can't automate these things. They will work only after the user goes into the submodule and checks out a branch.

### Case 2. Pulling Upstream Changes from the Project Remote

Someone just did the Case 1. procedure. At this point the other writers need to:

- Pull in the parent project changes
- run git submodule update because the previous pull did not update the submodules. Actually, this step is not currently possible in our Git plugin because  we don't have a git submodule update equivalent.

What we want to do on the short term to help with this use case is  to  add an option _"Update all submodules when pulling changes in the parent repository"_ . It will probably be selected by default and it will trigger a `git submodule update` after each Pull operation.

I agree that it would be nice to have more flexibility, like the selective update that you describe above, but we believe that most of the time people want to update all submodules.

Please let me know your thoughts on the matter. I'm especially curious to know if you use `.gitmodule` to specify a branch for your submodules.

**Comment by kirkilj** [comment]
I'm sorry I haven't chimed in earlier when you asked for submodule use cases.

Our most common use cases have the superproject using branches named after product release numbers, which are usually named after the year and month of a product release. Sometimes a submodule might also be using branches based on release numbers, but they most often use GIT TAGS instead, based on semver numbering. In this later case, they explicitly want to lock a dependency down to a specific commit, and not to a branch with a moving HEAD. In such a case, the writers on the super project want to know if any new tags have been added to the submodules' repos after the one currently reflected by the current submoduleS' references. If they saw a list of these new tags with date/author annotations, they could research it with their peers to see if it's safe to update the submodule reference to use one of the newer tags.

As in other dependency management tools, using "latest" might be fine for some use cases perhaps, but in the great majority of cases, a moving HEAD is risky business if compatibility cannot be assured. 

So it seems like there are 2 distinct dependency management methodologies, which could be chosen in a preference setting that could be tied to a specific project.

I'm still waiting on our Information Architecture team to weigh in, but I thought it might be useful for you to think about tag-based submodule workflows.

**Comment by AlexJitianu** [comment]
I'm trying to visualize how a user would work with the Git Client in such a environment. Let's assume this workflow:
1. We present a notification or a tooltip that says "There are N new tag(s) available for submodule SUBMODULE_NAME"
2. The user open SUBMODULE_NAME
Switching to a newer tag gets tricky. We don't have such an action so the only workaround that I can think of is to:
3. Move the submodule to a branch
4. Show history for the branch and look for the Tag between the commits of the branch
5. Use the _Reset {branchName} to this commit_ to move the submodule on the tag commit
I guess it would help if the _Git branch manager_ view would present the tags too.
